<a href="https://colab.research.google.com/github/qelmartinez/CPE310-IE22S1/blob/main/MARTINEZ_%7C_Getting_Started_With_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Section 1: Getting Started With Pandas

We will begin by introducing the `Series`, `DataFrame`, and `Index` classes, which are the basic building blocks of the pandas library, and showing how to work with them. By the end of this section, you will be able to create DataFrames and perform operations on them to inspect and filter the data.

## Anatomy of a DataFrame

A **DataFrame** is composed of one or more **Series**. The names of the **Series** form the column names, and the row labels form the **Index**.

In [8]:
import pandas as pd

meteorites = pd.read_csv('/content/Meteorite_Landings.csv', nrows=5)

*Source: [NASA's Open Data Portal](https://data.nasa.gov/Space-Science/Meteorite-Landings/gh4g-9sfh)*

In [9]:
meteorites

name   id nametype     recclass  mass (g)  fall  \
0    Aachen    1    Valid           L5        21  Fell   
1    Aarhus    2    Valid           H6       720  Fell   
2      Abee    6    Valid          EH4    107000  Fell   
3  Acapulco   10    Valid  Acapulcoite      1914  Fell   
4   Achiras  370    Valid           L6       780  Fell   

                     year    reclat    reclong           GeoLocation  
0  01/01/1880 12:00:00 AM  50.77500    6.08333     (50.775, 6.08333)  
1  01/01/1951 12:00:00 AM  56.18333   10.23333  (56.18333, 10.23333)  
2  01/01/1952 12:00:00 AM  54.21667 -113.00000    (54.21667, -113.0)  
3  01/01/1976 12:00:00 AM  16.88333  -99.90000     (16.88333, -99.9)  
4  01/01/1902 12:00:00 AM -33.16667  -64.95000   (-33.16667, -64.95)

#### Series:

In [ ]:
meteorites.name

0      Aachen
1      Aarhus
2        Abee
3    Acapulco
4     Achiras
Name: name, dtype: object

#### Columns:

In [ ]:
meteorites.columns #attribute NOT a function, ()

Index(['name', 'id', 'nametype', 'recclass', 'mass (g)', 'fall', 'year',
       'reclat', 'reclong', 'GeoLocation'],
      dtype='object')

#### Index:

In [ ]:
meteorites.index

RangeIndex(start=0, stop=5, step=1)

In [ ]:
sampleList = ['apple', 'orange', 'kiwi']

In [ ]:
sampleList[-1]

'kiwi'

## Creating DataFrames

We can create DataFrames from a variety of sources such as other Python objects, flat files, webscraping, and API requests. Here, we will see just a couple of examples, but be sure to check out [this page](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html) in the documentation for a complete list.

### Using a flat file

In [ ]:
import pandas as pd

meteorites = pd.read_csv('/content/meteorite_landings.csv')

*Tip: There are many parameters to this function to handle some initial processing while reading in the file &ndash; be sure check out the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html).*

### Using data from an API

Collect the data from [NASA's Open Data Portal](https://data.nasa.gov/Space-Science/Meteorite-Landings/gh4g-9sfh) using the Socrata Open Data API (SODA) with the `requests` library:

In [13]:
import requests

response = requests.get(
    'https://data.nasa.gov/resource/gh4g-9sfh.json',
    params={'$limit': 50_000}
)

if response.ok:
    payload = response.json()
else:
    print(f'Request was not successful and returned code: {response.status_code}.')
    payload = None

Request was not successful and returned code: 404.


Create the DataFrame with the resulting payload:

In [ ]:
import pandas as pd

df = pd.DataFrame(payload)
df.head(3)

*Tip: `df.to_csv('data.csv')` writes this data to a new file called `data.csv`.*

## Inspecting the data
Now that we have some data, we need to perform an initial inspection of it. This gives us information on what the data looks like, how many rows/columns there are, and how much data we have.

Let's inspect the `meteorites` data.

#### How many rows and columns are there?

In [ ]:
meteorites.shape

(45716, 10)

#### What are the column names?

In [ ]:
meteorites.columns

Index(['name', 'id', 'nametype', 'recclass', 'mass (g)', 'fall', 'year',
       'reclat', 'reclong', 'GeoLocation'],
      dtype='object')

#### What type of data does each column currently hold?

In [14]:
meteorites.dtypes

name            object
id               int64
nametype        object
recclass        object
mass (g)         int64
fall            object
year            object
reclat         float64
reclong        float64
GeoLocation     object
dtype: object

#### What does the data look like?

In [15]:
meteorites.head(15) #know first 5 rows; specify if you want higher

name   id nametype     recclass  mass (g)  fall  \
0    Aachen    1    Valid           L5        21  Fell   
1    Aarhus    2    Valid           H6       720  Fell   
2      Abee    6    Valid          EH4    107000  Fell   
3  Acapulco   10    Valid  Acapulcoite      1914  Fell   
4   Achiras  370    Valid           L6       780  Fell   

                     year    reclat    reclong           GeoLocation  
0  01/01/1880 12:00:00 AM  50.77500    6.08333     (50.775, 6.08333)  
1  01/01/1951 12:00:00 AM  56.18333   10.23333  (56.18333, 10.23333)  
2  01/01/1952 12:00:00 AM  54.21667 -113.00000    (54.21667, -113.0)  
3  01/01/1976 12:00:00 AM  16.88333  -99.90000     (16.88333, -99.9)  
4  01/01/1902 12:00:00 AM -33.16667  -64.95000   (-33.16667, -64.95)

Sometimes there may be extraneous data at the end of the file, so checking the bottom few rows is also important:

In [ ]:
meteorites.tail(10)

name     id nametype              recclass   mass (g)   fall  \
45706     Zerkaly  31354    Valid                    H5    16000.0  Found   
45707    Zhaoping  54609    Valid     Iron, IAB complex  2000000.0  Found   
45708    Zhigansk  30405    Valid           Iron, IIIAB   900000.0  Found   
45709  Zhongxiang  30406    Valid                  Iron   100000.0  Found   
45710  Zillah 001  31355    Valid                    L6     1475.0  Found   
45711  Zillah 002  31356    Valid               Eucrite      172.0  Found   
45712      Zinder  30409    Valid  Pallasite, ungrouped       46.0  Found   
45713        Zlin  30410    Valid                    H4        3.3  Found   
45714   Zubkovsky  31357    Valid                    L6     2167.0  Found   
45715  Zulu Queen  30414    Valid                  L3.7      200.0  Found   

                         year    reclat    reclong             GeoLocation  
45706  01/01/1956 12:00:00 AM  52.13333   81.96667    (52.13333, 81.96667)  
45707  01/01/1983 12:00:00 AM  24.23333  111.18333   (24.23333, 111.18333)  
45708  01/01/1966 12:00:00 AM  68.00000  128.30000           (68.0, 128.3)  
45709  01/01/1981 12:00:00 AM  31.20000  112.50000           (31.2, 112.5)  
45710  01/01/1990 12:00:00 AM  29.03700   17.01850       (29.037, 17.0185)  
45711  01/01/1990 12:00:00 AM  29.03700   17.01850       (29.037, 17.0185)  
45712  01/01/1999 12:00:00 AM  13.78333    8.96667     (13.78333, 8.96667)  
45713  01/01/1939 12:00:00 AM  49.25000   17.66667       (49.25, 17.66667)  
45714  01/01/2003 12:00:00 AM  49.78917   41.50460     (49.78917, 41.5046)  
45715  01/01/1976 12:00:00 AM  33.98333 -115.68333  (33.98333, -115.68333)

#### Get some information about the DataFrame

In [ ]:
meteorites.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45716 entries, 0 to 45715
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         45716 non-null  object 
 1   id           45716 non-null  int64  
 2   nametype     45716 non-null  object 
 3   recclass     45716 non-null  object 
 4   mass (g)     45585 non-null  float64
 5   fall         45716 non-null  object 
 6   year         45425 non-null  object 
 7   reclat       38401 non-null  float64
 8   reclong      38401 non-null  float64
 9   GeoLocation  38401 non-null  object 
dtypes: float64(3), int64(1), object(6)
memory usage: 3.5+ MB


## Extracting subsets

A crucial part of working with DataFrames is extracting subsets of the data: finding rows that meet a certain set of criteria, isolating columns/rows of interest, etc. After narrowing down our data, we are closer to discovering insights. This section will be the backbone of many analysis tasks.

#### Selecting columns

We can select columns as attributes if their names would be valid Python variables:

In [ ]:
names = meteorites.name

In [ ]:
names # Created a subset

0            Aachen
1            Aarhus
2              Abee
3          Acapulco
4           Achiras
            ...    
45711    Zillah 002
45712        Zinder
45713          Zlin
45714     Zubkovsky
45715    Zulu Queen
Name: name, Length: 45716, dtype: object

If they aren't, we have to select them as keys. However, we can select multiple columns at once this way:

In [16]:
x = meteorites[['name', 'mass (g)']]

#### Selecting rows

In [ ]:
meteorites[100:104:2]

name    id nametype       recclass  mass (g)  fall  \
100  Benton  5026    Valid            LL6    2840.0  Fell   
102  Béréba  5028    Valid  Eucrite-mmict   18000.0  Fell   

                       year  reclat  reclong      GeoLocation  
100  01/01/1949 12:00:00 AM   45.95   -67.55  (45.95, -67.55)  
102  01/01/1924 12:00:00 AM   11.65    -3.65   (11.65, -3.65)

#### Indexing

We use `iloc[]` to select rows and columns by their position:

In [18]:
meteorites.iloc[0:20, 2:5]

Empty DataFrame
Columns: [nametype, recclass, mass (g)]
Index: []

We use `loc[]` to select by name:

In [ ]:
meteorites.loc[100:104, 'mass (g)':'year']

mass (g)  fall                    year
100    2840.0  Fell  01/01/1949 12:00:00 AM
101     270.0  Fell  01/01/2008 12:00:00 AM
102   18000.0  Fell  01/01/1924 12:00:00 AM
103    1440.0  Fell  01/01/1811 12:00:00 AM
104     960.0  Fell  01/01/2004 12:00:00 AM

#### Filtering with Boolean masks

A **Boolean mask** is a array-like structure of Boolean values &ndash; it's a way to specify which rows/columns we want to select (`True`) and which we don't (`False`).

Here's an example of a Boolean mask for meteorites weighing more than 50 grams that were found on Earth (i.e., they were not observed falling):

In [ ]:
(meteorites['mass (g)'] > 50) & (meteorites.fall == 'Found')

0        False
1        False
2        False
3        False
4        False
         ...  
45711     True
45712    False
45713    False
45714     True
45715     True
Length: 45716, dtype: bool

**Important**: Take note of the syntax here. We surround each condition with parentheses, and we use bitwise operators (`&`, `|`, `~`) instead of logical operators (`and`, `or`, `not`).

We can use a Boolean mask to select the subset of meteorites weighing more than 1 million grams (1,000 kilograms or roughly 2,205 pounds) that were observed falling:

In [ ]:
meteorites[(meteorites['mass (g)'] > 1e6) & (meteorites.fall == 'Fell')]

name     id nametype    recclass    mass (g)  fall  \
29         Allende   2278    Valid         CV3   2000000.0  Fell   
419          Jilin  12171    Valid          H5   4000000.0  Fell   
506  Kunya-Urgench  12379    Valid          H5   1100000.0  Fell   
707  Norton County  17922    Valid     Aubrite   1100000.0  Fell   
920   Sikhote-Alin  23593    Valid  Iron, IIAB  23000000.0  Fell   

                       year    reclat    reclong             GeoLocation  
29   01/01/1969 12:00:00 AM  26.96667 -105.31667  (26.96667, -105.31667)  
419  01/01/1976 12:00:00 AM  44.05000  126.16667      (44.05, 126.16667)  
506  01/01/1998 12:00:00 AM  42.25000   59.20000           (42.25, 59.2)  
707  01/01/1948 12:00:00 AM  39.68333  -99.86667   (39.68333, -99.86667)  
920  01/01/1947 12:00:00 AM  46.16000  134.65333      (46.16, 134.65333)

*Tip: Boolean masks can be used with `loc[]` and `iloc[]`.*

An alternative to this is the `query()` method:

In [ ]:
meteorites.query("`mass (g)` > 1e6 and fall == 'Fell'")

name     id nametype    recclass    mass (g)  fall  \
29         Allende   2278    Valid         CV3   2000000.0  Fell   
419          Jilin  12171    Valid          H5   4000000.0  Fell   
506  Kunya-Urgench  12379    Valid          H5   1100000.0  Fell   
707  Norton County  17922    Valid     Aubrite   1100000.0  Fell   
920   Sikhote-Alin  23593    Valid  Iron, IIAB  23000000.0  Fell   

                       year    reclat    reclong             GeoLocation  
29   01/01/1969 12:00:00 AM  26.96667 -105.31667  (26.96667, -105.31667)  
419  01/01/1976 12:00:00 AM  44.05000  126.16667      (44.05, 126.16667)  
506  01/01/1998 12:00:00 AM  42.25000   59.20000           (42.25, 59.2)  
707  01/01/1948 12:00:00 AM  39.68333  -99.86667   (39.68333, -99.86667)  
920  01/01/1947 12:00:00 AM  46.16000  134.65333      (46.16, 134.65333)

*Tip: Here, we can use both logical operators and bitwise operators.*

## Calculating summary statistics

In the next section of this workshop, we will discuss data cleaning for a more meaningful analysis of our datasets; however, we can already extract some interesting insights from the `meteorites` data by calculating summary statistics.

#### How many of the meteorites were found versus observed falling?

In [ ]:
meteorites.fall.value_counts()

fall
Found    44609
Fell      1107
Name: count, dtype: int64

[The Meteoritical Society states](https://www.lpi.usra.edu/meteor/metbull.php?code=56147) that "relict meteorites are composed mostly of terrestrial minerals, but are thought to have once been meteorites." What proportion of the data are relict meteorites? Let's verify these were all found versus observed falling:

In [ ]:
meteorites.value_counts(subset=['nametype', 'fall'], normalize=True)

nametype  fall 
Valid     Found    0.974145
          Fell     0.024215
Relict    Found    0.001641
Name: proportion, dtype: float64

#### What was the mass of the average meterorite?

In [ ]:
meteorites['mass (g)'].mean()

np.float64(13278.078548601512)

**Important**: The mean isn't always the best measure of central tendency. If there are outliers in the distribution, the mean will be skewed. Here, the mean is being pulled higher by some very heavy meteorites &ndash; the distribution is [right-skewed](https://www.analyticsvidhya.com/blog/2020/07/what-is-skewness-statistics/).

Taking a look at some quantiles at the extremes of the distribution shows that the mean is between the 95th and 99th percentile of the distribution, so it isn't a good measure of central tendency here:

In [ ]:
meteorites['mass (g)'].quantile([0.01, 0.05, 0.5, 0.95, 0.99])

0.01        0.44
0.05        1.10
0.50       32.60
0.95     4000.00
0.99    50600.00
Name: mass (g), dtype: float64

A better measure in this case is the median (50th percentile), since it is robust to outliers:

In [ ]:
meteorites['mass (g)'].median()

32.6

#### What was the mass of the heaviest meteorite?

In [ ]:
meteorites['mass (g)'].max()

60000000.0

Let's extract the information on this meteorite:

In [ ]:
meteorites.loc[meteorites['mass (g)'].idxmax()]

name                             Hoba
id                              11890
nametype                        Valid
recclass                    Iron, IVB
mass (g)                   60000000.0
fall                            Found
year           01/01/1920 12:00:00 AM
reclat                      -19.58333
reclong                      17.91667
GeoLocation     (-19.58333, 17.91667)
Name: 16392, dtype: object

#### How many different types of meteorite classes are represented in this dataset?

In [ ]:
meteorites.recclass.nunique()

466

Some examples:

In [ ]:
meteorites.recclass.unique()[:14]

array(['L5', 'H6', 'EH4', 'Acapulcoite', 'L6', 'LL3-6', 'H5', 'L',
       'Diogenite-pm', 'Unknown', 'H4', 'H', 'Iron, IVA', 'CR2-an'],
      dtype=object)

*Note: All fields preceded with "rec" are the values recommended by The Meteoritical Society. Check out [this Wikipedia article](https://en.wikipedia.org/wiki/Meteorite_classification) for some information on meteorite classes.*

#### Get some summary statistics on the data itself
We can get common summary statistics for all columns at once. By default, this will only be numeric columns, but here, we will summarize everything together:

In [ ]:
meteorites.describe(include='all')

name            id nametype recclass      mass (g)   fall  \
count        45716  45716.000000    45716    45716  4.558500e+04  45716   
unique       45716           NaN        2      466           NaN      2   
top     Zulu Queen           NaN    Valid       L6           NaN  Found   
freq             1           NaN    45641     8285           NaN  44609   
mean           NaN  26889.735104      NaN      NaN  1.327808e+04    NaN   
std            NaN  16860.683030      NaN      NaN  5.749889e+05    NaN   
min            NaN      1.000000      NaN      NaN  0.000000e+00    NaN   
25%            NaN  12688.750000      NaN      NaN  7.200000e+00    NaN   
50%            NaN  24261.500000      NaN      NaN  3.260000e+01    NaN   
75%            NaN  40656.750000      NaN      NaN  2.026000e+02    NaN   
max            NaN  57458.000000      NaN      NaN  6.000000e+07    NaN   

                          year        reclat       reclong GeoLocation  
count                    45425  38401.000000  38401.000000       38401  
unique                     266           NaN           NaN       17100  
top     01/01/2003 12:00:00 AM           NaN           NaN  (0.0, 0.0)  
freq                      3323           NaN           NaN        6214  
mean                       NaN    -39.122580     61.074319         NaN  
std                        NaN     46.378511     80.647298         NaN  
min                        NaN    -87.366670   -165.433330         NaN  
25%                        NaN    -76.714240      0.000000         NaN  
50%                        NaN    -71.500000     35.666670         NaN  
75%                        NaN      0.000000    157.166670         NaN  
max                        NaN     81.166670    354.473330         NaN

**Important**: `NaN` values signify missing data. For instance, the `fall` column contains strings, so there is no value for `mean`; likewise, `mass (g)` is numeric, so we don't have entries for the categorical summary statistics (`unique`, `top`, `freq`).

#### Check out the documentation for more descriptive statistics:

- [Series](https://pandas.pydata.org/docs/reference/series.html#computations-descriptive-stats)
- [DataFrame](https://pandas.pydata.org/docs/reference/frame.html#computations-descriptive-stats)

# Exercises

A DataFrame is composed of one or more Series. The names of the Series form the column names, and the row labels form the Index.

In [33]:
import pandas as pd

yeltax = pd.read_csv('/content/2019_Yellow_Taxi_Trip_Data.csv')

In [29]:
yeltax

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           7.93           1                  N   
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           138           170             1         29.5      1      0.5   
1            11            26             1         10.5      1      0.5   
2           163           162             1          9.5      1      0.5   
3           170           163             1         13.0      1      0.5   
4           163           236             1         10.5      1      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        7.98          6.12                    0.3         47.90   
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

In [98]:
# Exercise 1.1 Create a DataFrame by reading in the 2019_Yellow_Taxi_Trip_Data.csv file. Examine the first 5 rows.
yeltax = pd.read_csv('/content/2019_Yellow_Taxi_Trip_Data.csv', nrows=5)

In [99]:
yeltax

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           7.93           1                  N   
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           138           170             1         29.5      1      0.5   
1            11            26             1         10.5      1      0.5   
2           163           162             1          9.5      1      0.5   
3           170           163             1         13.0      1      0.5   
4           163           236             1         10.5      1      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        7.98          6.12                    0.3         47.90   
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

**ANALYSIS**: When the code was executed, a table with 5 rows and 18 columns appeared. The columns were comprised of different categories related to taxi transactions, namely:vendorid,	tpep_pickup_datetime,	tpep_dropoff_datetime,	passenger_count, trip_distance, ratecodeid,	store_and_fwd_flag,	pulocationid,dolocationid, payment_type,	fare_amount,	extra, mta_tax,	tip_amount,	tolls_amount,	improvement_surcharge, total_amount, and	congestion_surcharge. This item was fairly the easiest, since we just copied the code from the initiating part.

In [102]:
# Exercise 1.2 Find the dimensions (number of rows and number of columns) in the data.
yeltax = pd.read_csv('/content/2019_Yellow_Taxi_Trip_Data.csv')
yeltax.shape

(10000, 18)

**ANALYSIS:** For this item, I rerun the initial code to ensure that all the rows will be counted. Comparing it on my initial trial that simply typed the "yeltax.shape" with the dimension of (5, 18), this reran code gave us (1000, 18) which is complete.

In [100]:
# Exercise 1.3 Using the data in the 2019_Yellow_Taxi_Trip_Data.csv file, calculate summary statistics for the fare_amount, tip_amount, tolls_amount, and total_amount columns.
yeltax[['fare_amount','tip_amount', 'tolls_amount', 'total_amount']].describe(include='all')

fare_amount  tip_amount  tolls_amount  total_amount
count     5.000000    5.000000      5.000000       5.00000
mean     14.600000    2.960000      1.224000      22.58400
std       8.429116    3.270505      2.736947      14.54879
min       9.500000    0.000000      0.000000      12.30000
25%      10.500000    0.500000      0.000000      15.30000
50%      10.500000    2.000000      0.000000      15.80000
75%      13.000000    4.320000      0.000000      21.62000
max      29.500000    7.980000      6.120000      47.90000

**ANALYSIS:** 5 columns which serves as the decriptors for the categories (fare_amount, tip_amount, tolls_amount,	total_amount) from the original data and 8 statistical summary that includes count, mean, standard deviation, quantile, and maximum range appeared as the function ".describe(include='all')" was used. Meanwhile, the values inside the []/square bracket served as the one that limits/calls for the specific values.


In [91]:
# Exercise 1.4 Isolate the fare_amount, tip_amount, tolls_amount, and total_amount for the longest trip by distance (trip_distance).
yeltax[['fare_amount', 'tip_amount', 'tolls_amount', 'total_amount']].loc[yeltax['trip_distance'].idxmax()]

fare_amount     176.00
tip_amount       18.29
tolls_amount      6.12
total_amount    201.21
Name: 8338, dtype: float64

**ANALYSIS:** Just like the code above, I used []/square bracket to call on the specific values and create rows and then used "*.loc[yeltax['trip_distance'].idxmax()*" to just call on the data with highest data. According to the table that appeared, id 8338 has the highest distance with a fare amount of 176, a tip of 18.29, and toll amont of 6.12, totalling 201.21.